ledger/trie/trie_test.go

```diff
@@ -0,0 +1,278 @@
+	trie := trie.NewEmptyTrie(store)
+
+	path := utils.PathByUint16LeftPadded(0)
+	payload := utils.LightPayload(11, 12345)
+
+	trie.Insert(path, payload)
+
+	expectedRootHashHex := "b30c99cc3e027a6ff463876c638041b1c55316ed935f1b3699e52a2c3e3eaaab"
+
+	got := trie.RootHash()
+	require.Equal(t, expectedRootHashHex, hex.EncodeToString(got[:]))
+}
+
+func TestTrie_InsertRightRegister(t *testing.T) {
+	store := mocks.BaselineStore()
+	trie := trie.NewEmptyTrie(store)
+
+	var path ledger.Path
+	for i := 0; i < len(path); i++ {
+		path[i] = uint8(255)
+	}
+	payload := utils.LightPayload(12346, 54321)
+
+	trie.Insert(path, payload)
+
+	expectedRootHashHex := "4313d22bcabbf21b1cfb833d38f1921f06a91e7198a6672bc68fa24eaaa1a961"
+
+	got := trie.RootHash()
+	require.Equal(t, expectedRootHashHex, hex.EncodeToString(got[:]))
+}
+
+func TestTrie_InsertMiddleRegister(t *testing.T) {
+	store := mocks.BaselineStore()
+	trie := trie.NewEmptyTrie(store)
+
+	path := utils.PathByUint16LeftPadded(56809)
+	payload := utils.LightPayload(12346, 59656)
+
+	trie.Insert(path, payload)
+
+	expectedRootHashHex := "4a29dad0b7ae091a1f035955e0c9aab0692b412f60ae83290b6290d4bf3eb296"
+
+	got := trie.RootHash()
+	require.Equal(t, expectedRootHashHex, hex.EncodeToString(got[:]))
+}
+
+func TestTrie_InsertManyRegisters(t *testing.T) {
+	dir, err := os.MkdirTemp("", "")
+	require.NoError(t, err)
+	defer os.RemoveAll(dir)
+
+	store, err := store.NewStore(zerolog.New(io.Discard), 4*1000*1000, dir)
+	require.NoError(t, err)
+
+	trie := trie.NewEmptyTrie(store)
+
+	paths, payloads := helpers.SampleRandomRegisterWrites(helpers.NewGenerator(), 12001)
+
+	for i := range paths {
+		trie.Insert(paths[i], &payloads[i])
+	}
+
+	got := trie.RootHash()
+	expectedRootHashHex := "74f748dbe563bb5819d6c09a34362a048531fd9647b4b2ea0b6ff43f200198aa"
+	assert.Equal(t, expectedRootHashHex, hex.EncodeToString(got[:]))
+}
+
+// FIXME: Make a mock store which can substitute the store and preserve functionality for tests.
+
+func TestTrie_InsertFullTrie(t *testing.T) {
+	dir, err := os.MkdirTemp("", "")
+	require.NoError(t, err)
+	defer os.RemoveAll(dir)
+
+	store, err := store.NewStore(zerolog.New(io.Discard), 4*1000*1000, dir)
+	require.NoError(t, err)
+
+	trie := trie.NewEmptyTrie(store)
+
+	numberRegisters := 65536
+	rng := helpers.NewGenerator()
+	paths := make([]ledger.Path, 0, numberRegisters)
+	payloads := make([]*ledger.Payload, 0, numberRegisters)
+	for i := 0; i < numberRegisters; i++ {
+		paths = append(paths, utils.PathByUint16LeftPadded(uint16(i)))
+		temp := rng.Next()
+		payload := utils.LightPayload(temp, temp)
+		payloads = append(payloads, payload)
+	}
+
+	for i := range paths {
+		trie.Insert(paths[i], payloads[i])
+	}
+
+	got := trie.RootHash()
+	expectedRootHashHex := "6b3a48d672744f5586c571c47eae32d7a4a3549c1d4fa51a0acfd7b720471de9"
+	assert.Equal(t, expectedRootHashHex, hex.EncodeToString(got[:]))
+}
+
+func TestTrie_InsertManyTimes(t *testing.T) {
+	expectedRootHashes := []string{
```

Comment:
Hardcoded hashes and keys can be separated in a helper file for testing

Reply:
I'd rather put them as consts at the top of the test file if it's ok with you. Having a separate file is overkill for this case IMO

Reply:
If there are not so many repeating code blocks currently, thats fine yes

Reply:
Yes, sure, but if you create a helper file later as you said in the previous comment, they can be there as well